File: src/trafficpulse/app/routers/evidence.py
"""Evidence endpoints: the manifest, its rendered artifacts, and its package.

The manifest endpoint (H7A) now serves the **composed** record: the write-once
manifest the run persisted, with any rendered-artifact references merged in at read
time (H14). The persisted file is never modified, so a repository written before
H14 is served exactly as it always was.

Two endpoints join it. ``/artifacts/{kind}`` serves the bytes of one rendered
evidence frame -- the only way a client should obtain evidence pixels, because it is
the only source whose provenance the backend can state. ``/package`` serves the
deterministic ZIP an analyst downloads.
"""

from __future__ import annotations

from fastapi import APIRouter, Response

from ...contracts import EvidenceManifest
from ...contracts.enums import ArtifactKind
from ..dependencies import EvidenceServiceDep
from ..models import ErrorResponse

router = APIRouter(tags=["evidence"])


@router.get(
    "/api/evidence/{event_id}",
    response_model=EvidenceManifest,
    summary="Evidence manifest for an event",
    description="Return the evidence manifest for a confirmed event: the record the "
    "run persisted, with references to any rendered artifacts merged in. A frame "
    "reference carrying a `sha256` is fetchable through the artifacts endpoint; one "
    "without is a pre-render reference to a frame that was never materialised.",
    responses={404: {"model": ErrorResponse, "description": "Unknown event id"}},
)
def get_evidence(event_id: str, evidence: EvidenceServiceDep) -> EvidenceManifest:
    return evidence.get(event_id)


@router.get(
    "/api/evidence/{event_id}/artifacts/{kind}",
    summary="Rendered evidence artifact",
    description="Serve one rendered evidence frame (`before_frame`, `trigger_frame`, "
    "or `after_frame`) as an image. These are the frames the engine actually picked, "
    "drawn with the same overlay renderer as the annotated video, and verified "
    "against the SHA-256 the manifest records before they are served. 404 when "
    "nothing was rendered for that slot.",
    response_class=Response,
    responses={
        200: {"content": {"image/*": {}}, "description": "The rendered frame."},
        404: {
            "model": ErrorResponse,
            "description": "Unknown event, or no rendered artifact of that kind",
        },
    },
)
def get_evidence_artifact(
    event_id: str, kind: ArtifactKind, evidence: EvidenceServiceDep
) -> Response:
    data, media_type = evidence.artifact(event_id, kind)
    # Artifacts are content-addressed and write-once, so their bytes can never change
    # under a locator -- they are safe to cache immutably.
    return Response(
        content=data,
        media_type=media_type,
        headers={"Cache-Control": "public, max-age=31536000, immutable"},
    )


@router.get(
    "/api/evidence/{event_id}/package",
    summary="Evidence package (ZIP)",
    description="Download the complete evidence package for one event: the frozen "
    "confirmed event, the served manifest, and every rendered frame. Deterministic "
    "-- the same event and artifacts always produce byte-identical bytes -- so the "
    "archive's own hash is a property of the evidence rather than of the download.",
    response_class=Response,
    responses={
        200: {"content": {"application/zip": {}}, "description": "The evidence package."},
        404: {"model": ErrorResponse, "description": "Unknown event id"},
    },
)
def get_evidence_package(event_id: str, evidence: EvidenceServiceDep) -> Response:
    data, filename = evidence.package(event_id)
    return Response(
        content=data,
        media_type="application/zip",
        headers={"Content-Disposition": f'attachment; filename="{filename}"'},
    )

File: frontend/src/lib/evidence.ts
import { endpoints } from '@/api/endpoints';
import { type ArtifactReference, type EvidenceManifest } from '@/api/types';
import { env } from '@/lib/env';

/**
 * Evidence artifact selection (H14) — pure, no React, no network.
 *
 * The frontend does not decide which frames are evidence. Before H14 it did: the
 * viewer seeked the playing video to `trigger ± 1.5s` and called the result an
 * evidence frame, while the backend had independently picked the frames the engine
 * actually processed at its own configured margins. The two could disagree, and
 * nothing reconciled them.
 *
 * Now the manifest is the authority. A frame is evidence if, and only if, the
 * backend rendered it and the served manifest references it — which is exactly the
 * references carrying a `sha256`, since a content address is what a rendered
 * artifact has and a pre-render placeholder does not.
 */

/** The frame slots an evidence manifest can carry, in presentation order. */
export const EVIDENCE_FRAME_KINDS = ['before_frame', 'trigger_frame', 'after_frame'] as const;

export type EvidenceFrameKind = (typeof EVIDENCE_FRAME_KINDS)[number];

const FRAME_LABELS: Record<EvidenceFrameKind, string> = {
  before_frame: 'Before',
  trigger_frame: 'Trigger',
  after_frame: 'After',
};

export interface EvidenceFrame {
  kind: EvidenceFrameKind;
  label: string;
  /** The manifest reference this frame was resolved from. */
  artifact: ArtifactReference;
  /** Where the backend serves the rendered image. */
  url: string;
}

/**
 * Whether a reference points at bytes the backend actually rendered.
 *
 * A reference without a `sha256` names a source frame the engine picked but nobody
 * materialised (every manifest written before H14 is this shape). Requesting it
 * would 404, so it is not offered as a frame.
 */
export function isRenderedArtifact(
  artifact: ArtifactReference | null | undefined,
): artifact is ArtifactReference {
  return Boolean(artifact && artifact.sha256);
}

/** The URL the backend serves one rendered evidence frame from. */
export function evidenceArtifactUrl(eventId: string, kind: EvidenceFrameKind): string {
  return `${env.apiBaseUrl}${endpoints.evidenceArtifact(eventId, kind)}`;
}

/** The URL the complete evidence package downloads from. */
export function evidencePackageUrl(eventId: string): string {
  return `${env.apiBaseUrl}${endpoints.evidencePackage(eventId)}`;
}

/**
 * The rendered evidence frames for an event, in before → trigger → after order.
 *
 * Empty when nothing was rendered — an event from a pre-H14 repository, or one whose
 * render failed. The caller shows the metadata it does have rather than inventing
 * pixels, which is the whole point of the change.
 */
export function evidenceFrames(
  eventId: string,
  manifest: EvidenceManifest | undefined | null,
): EvidenceFrame[] {
  if (!manifest) return [];
  const frames: EvidenceFrame[] = [];
  for (const kind of EVIDENCE_FRAME_KINDS) {
    const artifact = manifest[kind];
    if (!isRenderedArtifact(artifact)) continue;
    frames.push({
      kind,
      label: FRAME_LABELS[kind],
      artifact,
      url: evidenceArtifactUrl(eventId, kind),
    });
  }
  return frames;
}

/** Every artifact reference on a manifest, for the viewer's artifact table. */
export function manifestArtifacts(
  manifest: EvidenceManifest | undefined | null,
): Array<{ label: string; artifact: ArtifactReference | null }> {
  if (!manifest) return [];
  return [
    { label: 'Before frame', artifact: manifest.before_frame },
    { label: 'Trigger frame', artifact: manifest.trigger_frame },
    { label: 'After frame', artifact: manifest.after_frame },
    { label: 'Clip', artifact: manifest.clip },
    { label: 'Trajectory', artifact: manifest.trajectory },
    { label: 'Plate crop', artifact: manifest.plate_crop },
  ];
}

/** A short, readable form of an artifact's content address (or a dash). */
export function shortDigest(artifact: ArtifactReference | null | undefined): string {
  if (!artifact?.sha256) return '—';
  return artifact.sha256.slice(0, 12);
}
